lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -3329,7 +3329,17 @@ private function isIdentifierEquals($entity1, $entity2)
             ? $this->entityIdentifiers[$oid2]
             : $this->identifierFlattener->flattenIdentifier($class, $class->getIdentifierValues($entity2));
 
-        return $id1 === $id2 || implode(' ', $id1) === implode(' ', $id2);
+        return $id1 === $id2 || $this->hashIdentifier($id1) === $this->hashIdentifier($id2);
+    }
+
+    /**
+     * @param array $identifier
+     *
+     * @return string
+     */
+    private function hashIdentifier(array $identifier)
+    {
+        return serialize(array_values(array_map('strval', $identifier)));
```

Comment:
Keep in mind that Asmir worked on a similar issue a few years ago and we didn't accept because of the 2% performance impact... We might take a look at his approach too to verify all issues got solved. =)